File: offer/03.数组中重复的数字.js
/**
 * 输入：
 * [2, 3, 1, 0, 2, 5, 3]
 * 输出：2 或 3
 */

// function findRepeatNumber(nums) {
//   const map = {};
//   let res;
//   for (let num of nums) {
//     if (map[num] === undefined) {
//       map[num] = true;
//     } else {
//       res = num;
//     }
//   }
//   return res;
// }

function findRepeatNumber(nums) {
  const map = {};
  let res;
  for (let num of nums) {
    if (map[num]) {
      res = num;
      break;
    }
    map[num] = true;
  }
  return res;
}

console.log(findRepeatNumber([2, 3, 1, 0, 2, 5, 3]));

File: offer/03.数组中重复的数字.go
func findRepeatNumber(nums []int) int {
    obj := make(map[int]bool)
    var res int
    for _, num := range nums {
        if _, ok := obj[num]; !ok {
            obj[num] = true
        } else {
            res = num
        }
    }
    return res
}

func findRepeatNumber(nums []int) int {
    obj := make(map[int]bool)
    var res int
    for _, num := range nums {
        if obj[num] {
            res = num
            break
        }
        obj[num] = true
    }
    return res
}